lib/puppet_x/puppetlabs/powershell/powershell_manager.rb

```diff
@@ -200,58 +229,115 @@ def self.wait_on(wait_object, timeout_ms = 50)
+
+        # "#{msg.length}\n" + msg
+      end
+
+      def write_pipe(input)
+        @pipe.syswrite(input)
+        @pipe.flush()
       end
 
       def drain_pipe(pipe, iterations = 10)
         output = []
         0.upto(iterations) do
           break if !self.class.is_readable?(pipe, 0.1)
-          l = pipe.gets
+          # require 'pry'; binding.pry
+          l = pipe.readpartial(4096)
           Puppet.debug "#{Time.now} PIPE> #{l}"
           output << l
         end
         output
       end
 
-      def read_stdout(output_ready_event, wait_interval_ms = 50)
+      def read_pipe(output_ready_event, wait_interval_ms = 50)
         output = []
         errors = []
         waited = 0
 
         # drain the pipe while waiting for the event signal
+        # but prevent an infinite loop by validating the process, pipes, etc
         while WAIT_TIMEOUT == self.class.wait_on(output_ready_event, wait_interval_ms)
+          # Ruby will gleefully allow trying to read stdout / stderr that are
+          # no longer connected to a live process, so therefore check the
+          # liveness here and raise the broken pipe error that Ruby would usually
+          raise Errno::EPIPE if !alive?
           # TODO: While this does ensure that both pipes have been
           # drained it can block on either longer than necessary or
           # deadlock waiting for one or the other to finish. The correct
           # way to deal with this is to drain each pipe from seperate threads
           # but time ran on in this implementation and this will be addressed soon
-          output << drain_pipe(@stdout)
+          output << drain_pipe(@pipe)
           errors << drain_pipe(@stderr)
           waited += wait_interval_ms
         end
 
         Puppet.debug "Waited #{waited} total milliseconds."
 
         # once signaled, ensure everything has been drained
-        output << drain_pipe(@stdout, 1000)
+        output << drain_pipe(@pipe, 1000)
         errors << drain_pipe(@stderr, 1000)
 
         errors = errors.reject { |e| e.empty? }
+                  .map { |arr| arr.map { |e| e.force_encoding(Encoding::UTF_8) }}
 
-        return output.join(''), errors
-      rescue => e
-        msg = "Error reading PIPE: #{e}"
-        raise Puppet::Util::Windows::Error.new(msg)
+        return output.join('').force_encoding(Encoding::UTF_8), errors
       end
 
       def exec_read_result(powershell_code, output_ready_event)
-        write_stdin(powershell_code)
-        read_stdout(output_ready_event)
+        write_pipe(pipe_command(:execute))
+        write_pipe(pipe_data(powershell_code))
+        read_pipe(output_ready_event)
+      rescue EOFError => eoferror # a specific type of IOError
+        Puppet.warning [eoferror.inspect, eoferror.backtrace].join("\n")
+        # require 'pry'; binding.pry
+        # pipes are closed (could be as a result of tearing down process)
+        @usable = false
+        return nil, [eoferror.inspect]
```

Comment:
As per conversation. Perhaps change the warning to "Powershell Manager terminated unexpectedly" and add the backtrace to Debug.